### ISSUE TITLE
rc0: Bundled vega5 extension uses the wrong Vega-Lite version

### ISSUE THREAD

**Issue by jakevdp** [jupyterlab/jupyterlab#6709]
It should be Vega-Lite 3.3, but it is Vega-Lite 2.7

To reproduce in a clean environment:
```
$ pip install --pre juptyterlab altair vega_datasets
$ jupyter lab
```
Open a new notebook and run
```python
import altair as alt

# load a simple dataset as a pandas DataFrame
from vega_datasets import data
cars = data.cars()

alt.Chart(cars).mark_point().encode(
    x='Horsepower',
    y='Miles_per_Gallon',
    color='Origin',
).interactive()
```
Open the browser console and observe the warning:
```
The input spec uses vega-lite v3.3.0, but the current version of Vega-Lite is 2.7.0.
```

**Comment by jakevdp** [comment]
More direct way to see the error:
```
from IPython.display import display

display({
    "application/vnd.vegalite.v3+json": {
        "$schema": "https://vega.github.io/schema/vega-lite/v3.json",
        "mark": "errorbar",
    }
}, raw=True)
```
```
Javascript Error: Invalid mark type "errorbar"
```
errorbar is a valid mark in Vega-Lite 3.

**Comment by domoritz** [comment]
@saulshanabrook Probably knows best what's going on here. The versions in the `package.json`s look right to me.

**Comment by telamonian** [comment]
I'm able to reproduce your error from a clean build. It looks like there a version snarl amongst the dependencies of `@jupyterlab/vega5-extension`, which would be my guess for the source of the error. The Jlab package `vega5-extension` has a development dependency on both:

- `"vega-embed": "^4.2.0",`
- `"vega-lite": "^3.3.0",`

In turn, `vega-embed` has a dev dependency on:

- `"vega-lite": "^3.0.0"`

The truly weird thing is the `vega-lite: ^3.0.0` dependency of `vega-embed` seems to be resolving to `2.7.0`. There are many better options, including an actual `3.0.0` release.

In turn, the cause of the weird version resolution seems to be that the `vega-lite` folks [just released a maintanaince build of `2.7.0`](https://github.com/vega/vega-lite/releases). That means that the two most recent releases of `vega-lite` are `2.7.0` and `3.3.0`. `vega-embed`'s `^3.0.0` dependency ends up like a coin on it's edge in between the two release versions, and ends up falling inconveniently to the lower version. The `2.7.0` release was 5 days ago, so `@jupyterlab/vega5-extension` was probably working fine before then.

If this is indeed the problem, then I suppose the fix would be to ask the `vega-embed` people to bump their version of `vega-lite`. I've also never heard of a caret matching to a lower version, so it's possible that this is actually a bug in `yarn` or `npm`.

**Comment by domoritz** [comment]
We made the 2.7 release specifically for jupyterlab. I'm surprised it gets priority over the 3.3 release by the version resolution.

**Comment by telamonian** [comment]
I manually bumped the `vega-lite` dep in `vega-embed` to `^3.3.0` and rebuilt. Sadly I'm still getting a javascript error, but at least it's a different one:

```
Javascript Error: Cannot read property 'x' of undefined
```

Progress! Either of you know what that error is about?

**Comment by jakevdp** [comment]
Try the first snippet instead - the second one is incomplete.

**Comment by telamonian** [comment]
> Try the first snippet instead - the second one is incomplete.

Yep, that first snippet works now:

<img width="665" alt="image" src="https://user-images.githubusercontent.com/2263641/60191635-a6f5e480-9802-11e9-8dd6-fa61904827fc.png">

so I was definitely right that the `v2.7.0` version of `vega-lite` that's currently shipping with `vega-embed` is to blame for the original issue. 

Not sure how to fix the problem on Jlab's end; it looks like someone might have some ideas, though

**Comment by domoritz** [comment]
> Yep, that first snippet works now with my lightly hacked copy of @jupyterlab/vega5-extension:

Can you confirm that the javascript console doesn't show an error such as "The input spec uses vega-lite v3.3.0, but the current version of Vega-Lite is 2.7.0."?

I wonder why npm installs 2.7 just because we released it after 3.3. Maybe @kanitw and I should make another 3.x release.

**Comment by saulshanabrook** [comment]
I am able to reproduce this locally. @telamonian is correct that the issue is that the `vega-embed` inside of the `@jupyterlab/vega5-extension` pulls in a dependency of `vega-lite` of version 2.7.0. It is alongside the correct version of `vega-lite`, but it pulls its own version because of the [`nohoist` directive we added](https://github.com/jupyterlab/jupyterlab/blob/e486e636d26bbfd3b245deb2f1cc024a31f80ace/package.json#L19-L26) so that each package can have its own version.


Curiously, if you run `yarn why vega-lite` it incorrectly doesn't show this version inside `vega-embed`. I am going to try to fix this by changing our `nohoist` patterns, so that the `vega-embed` doesn't have another `vega-lite` version inside of it.

**Comment by saulshanabrook** [comment]
I think we might be hitting some version of https://github.com/yarnpkg/yarn/issues/5520, I am still playing with it.

**Comment by telamonian** [comment]
@domoritz My console is clear. Making a new `3.x` release of `vega-lite` seems like a good thing to try (presumably the version mismatch may cause issues outside of just Jupyterlab). As to whether that will fix the problem, it seems like it's 50-50. 

Honestly, the version matching behavior of `^3.0.0` just seems bizarre here. Maybe it was only checking the last two releases, or maybe `^3.0.0` has some special behavior where it's able to match either `2.x.x` or `3.x.x`.

**Comment by jakevdp** [comment]
Sorry for the confusing test case above. Here is a short snippet that will produce a valid chart if vega-lite 3.X is used, and result in an error if vega-lite 2.X is used:
```python
import altair as alt
from vega_datasets import data

alt.Chart(data.barley()).mark_errorbar().encode(
    y='site:O',
    x='yield:Q'
)
```

**Comment by telamonian** [comment]
That one works too:

![image](https://user-images.githubusercontent.com/2263641/60197288-1d97df80-980d-11e9-97a7-7ba33ec51bb2.png)

**Comment by saulshanabrook** [comment]
> @saulshanabrook Out of curiosity, is it possible that the strange version matching behavior is due to something that Jlab is doing during its initial build? Or is that all handled by yarn?

No it's all due to yarn. I believe it is a bug in the yarn `nohoist` combination with `peerDependencies`.

It might not be related to the `devDependency` of `vega-embed`, since that shouldn't be used when installing the package.

I was able to work around this by creating two separate packages, that each have their own `node_modules` that are outside of our workspaces. Then I was able to install the right versions in each of those, without any of the `nohoist` config, and now `vega-embed` seems to correctly use the version of `vega-lite` next to it, instead of installing a new one inside of it https://github.com/jupyterlab/jupyterlab/pull/6714
